django/contrib/contenttypes/fields.py

```diff
@@ -236,21 +218,29 @@ def __get__(self, instance, cls=None):
         if instance is None:
             return self
 
+        # Make sure to use ContentType.objects.get_for_id() to ensure that
+        # lookups are cached (see ticket #5570). This takes more code than
+        # the naive ``getattr(instance, self.ct_field)``, but has better
+        # performance when dealing with GFKs in loops and such.
+        f = self.model._meta.get_field(self.ct_field)
+        ct_id = getattr(instance, f.get_attname(), None)
+        pk_val = getattr(instance, self.fk_field)
+
         try:
-            return getattr(instance, self.cache_attr)
+            rel_obj = getattr(instance, self.cache_attr)
+            # If the cached instance does not match the raw content type
+            # or primary key, force a reload.
+            if rel_obj and ct_id != self.get_content_type(obj=rel_obj, using=instance._state.db).id:
+                raise AttributeError
+            if rel_obj._meta.pk.to_python(pk_val) != rel_obj._get_pk_val():
```

Comment:
I'm not too happy about calling `get_content_type` and `to_python` inside this `try:` block, but it was the easiest thing to do. On the other hand, I don't think the chances of accidentally catching a stray `AttributeError` from either of the two calls are high enough to warrant the amount of code that would be needed to ensure it does not happen.